src/mt_optimize.jl

```diff
@@ -0,0 +1,142 @@
+@metatheory_init ()
+
+Metatheory.iscall(x::Type{<:Symbolic}) = true
+
+using SymbolicUtils.Rewriters
+
+"""
+Converts `Mul` and `Add` to `Term`.
+"""
+function toterm(t::Mul{T}) where T
+    args = []
+    push!(args, t.coeff)
+    for (k, deg) in t.dict
+        push!(args, deg == 1 ? k : Term{T}(^, [k, deg]))
+    end
+    Term{T}(*, args)
+end
+
+function toterm(t::Add{T}) where T
+    args = []
+    for (k, coeff) in t.dict
+        push!(args, coeff == 1 ? k : Term{T}(*, [coeff, k]))
+    end
+    Term{T}(+, args)
+end
+
+function toterm(t::Pow{T}) where T
+    Term{T}(^, [t.base, t.exp])
+end
+
+toterm(t) = t
+
+"""
+Binarizes `Term`s with n-ary operations
+"""
+function unflatten(t::Symbolic{T}) where{T}
+    # TODO change to isterm after PR
+    if SymbolicUtils.isterm(t)
+        f = gethead(t)
+        if f == (+) || f == (*) || f == (-)  # check out for other binary ops TODO
+            a = getargs(t)
+            return foldl((x,y) -> Term{T}(f, [x, y]), a)
+        end
+    end
+    return t
+end
+
+unflatten(t) = t
+
+function preprocess(t)
+    Chain([Postwalk(toterm), Postwalk(unflatten)])(t) 
+end
+
+
+
+"""
+Equational rewrite rules for optimizing expressions
+"""
+opt_theory = @methodtheory begin
+    a + b == b + a
+    a * b == b * a
+    a * x + a * y == a*(x+y)
+    -1 * a == -a
+    a + (-1 * b) == a - b
+    x^-1 == 1/x 
+    1/x * a == a/x
+    # fraction rules 
+    # (a/b) + (c/b) => (a+c)/b
+    # trig functions
+    sin(x)/cos(x) == tan(x)
+    cos(x)/sin(x) == cot(x)
+    sin(x)^2 + cos(x)^2 => 1
+    sin(2a) == 2sin(a)cos(a)
+end
+
+
+"""
+Approximation of costs of operators in number 
+of CPU cycles required for the numerical computation
+
+See 
+ * https://latkin.org/blog/2014/11/09/a-simple-benchmark-of-various-math-operations/
+ * https://streamhpc.com/blog/2012-07-16/how-expensive-is-an-operation-on-a-cpu/
+ * https://github.com/triscale-innov/GFlops.jl
+"""
+const op_costs = Dict(
+    (+)     => 1,
+    (-)     => 1,
+    abs     => 2,
+    (*)     => 3,
+    exp     => 18,
+    (/)     => 24,
+    (^)     => 100,
+    log1p   => 124,
+    deg2rad => 125,
+    rad2deg => 125,
+    acos    => 127,
+    asind   => 128,
+    acsch   => 133,
+    sin     => 134,
+    cos     => 134,
+    atan    => 135,
+    tan     => 156,
+)
```

Comment:
@ChrisElrod let us know if there is a better reference for assigning some kind of cost to these (and other) primitive functions...

Reply:
It'd be great to create a library for this.

A recent paper though showed that with an extremely detailed CPU model, you can model performance quite accurately. But failing that, their extremely simple baseline model which didn't consider port usage or anything like that at all performed about as well as -- often better than -- tools like LLVM-MCA.
So it probably makes more sense to go simple, and for function like `cos`, count how many instructions they consist of.
It gets more complicated when you have branches inside of them of course. There, I'd probably take the maximum among paths (unless it's rare and much slower).

I think these functions are faster relative to `+` than indicated in that table.

Reply:
But I'd really like a library that can do things like picking between variants of instructions/functions that hit ports differently.
Like balancing divisions vs approximate reciprocals + Newton Raphson.
Would also be useful for decisions like how much to unroll.

The paper I mentioned only did Intel CPUs, meaning no AMD, ARM, Apple, etc.